packages/kas/src/nodes.ts

```diff
@@ -69,22 +54,82 @@ var randomFloat = function (min, max) {
 var ITERATIONS = 12;
 var TOLERANCE = 9; // decimal places
 
+function partition<T, V extends _.Collection<T>>(
+    list: V,
+    iteratee: _.CollectionIterator<T, boolean>,
+    context?: any,
+): [_.TypeOfCollection<V>[], _.TypeOfCollection<V>[]] {
+    const a: _.TypeOfCollection<V>[] = [];
+    const b: _.TypeOfCollection<V>[] = [];
+    _.forEach(list, (elem, key, ctx) => {
+        if (iteratee(elem, key, ctx)) {
+            a.push(elem);
+        } else {
+            b.push(elem);
+        }
+    });
+    return [a, b];
+}
+
+function isExpr(arg: string | Expr): arg is Expr {
+    return arg instanceof Expr;
+}
+
+function isRational(arg: Expr): arg is Rational {
+    return arg instanceof Rational;
+}
+
+function getFactors(expr: Expr): Expr[] {
+    if (expr instanceof Mul) {
+        return expr.terms;
+    } else {
+        return [expr];
+    }
+}
+
+type Hints = {
+    parens?: boolean;
+    divide?: boolean;
+    root?: boolean;
+    subtract?: boolean;
+    fraction?: boolean;
+    entered?: boolean;
+    negate?: boolean;
+    open?: boolean;
+};
+
+type CollectOptions = {
+    preciseFloats?: boolean;
+    once?: boolean;
+};
+
 /* abstract base expression node */
 class Expr {
+    hints: Hints;
+
     constructor() {
         this.hints = {
             parens: false,
         };
     }
 
     // this node's immediate constructor
-    func() {
-        abstract();
-    }
+    // TODO(kevinb): Write a proper type for this
+    func: any = (...args: any[]) => {
+        throw new Error(
+            "Abstract method - must override for expr: " +
+                // eslint-disable-next-line @babel/no-invalid-this
+                this.print(),
+        );
+    };
 
     // an array of the arguments to this node's immediate constructor
-    args() {
-        return abstract();
+    args(): (number | string | Expr)[] {
+        throw new Error(
+            "Abstract method - must override for expr: " +
+                // eslint-disable-next-line @babel/no-invalid-this
+                this.print(),
+        );
```

Comment:
Instead of throwing in abstract methods, I think we can use TypeScript to mark the method as abstract, e.g.
```ts
abstract args(): (number | string | Expr)[];
```